shoop/admin/modules/orders/views/create.py

```diff
@@ -207,45 +207,45 @@ def handle_customer_data(self, request):
+            "orderIdentifier": order.identifier,
+            "url": reverse("shoop_admin:order.list")
+        })
+
     def handle_source_data(self, request):
-        try:
-            state = json.loads(request.body.decode("utf-8"))["state"]
-            source = create_source_from_state(state, creator=request.user)
-            # Calculate final lines for confirmation
-            source.calculate_taxes(force_recalculate=True)
-            return {
-                "customerId": source.customer.id,
-                "taxfulTotal": format_money(source.taxful_total_price.amount),
-                "taxlessTotal": format_money(source.taxless_total_price.amount),
-                "totalDiscountAmount": format_money(source.total_discount.amount),
-                "orderLines": [encode_line(line) for line in source.get_final_lines(with_taxes=True)],
-                "billingAddress": source.billing_address.as_string_list() if source.billing_address else None,
-                "shippingAddress": source.shipping_address.as_string_list() if source.shipping_address else None,
-            }
-        except Exception as exc:
-            message = _("Could not proceed with order:")
-            if isinstance(exc, ValidationError):  # pragma: no branch
-                message += "\n" + "\n".join(force_text(err) for err in exc.messages)
-            else:  # pragma: no branch
-                message += " " + str(exc)
-            return JsonResponse({"success": False, "errorMessage": message}, status=400)
-
-    @atomic
+        return _handle_or_return_error(self._handle_source_data, request, _("Could not proceed with order:"))
+
     def handle_create(self, request):
-        try:
-            state = json.loads(request.body.decode("utf-8"))["state"]
-            order = create_order_from_state(state, creator=request.user)
-            messages.success(request, _("Order %(identifier)s created.") % vars(order))
-            return JsonResponse({
-                "success": True,
-                "orderIdentifier": order.identifier,
-                "url": reverse("shoop_admin:order.list")
-            })
-        except Exception as exc:
-            message = _("Could not create order:")
-            if isinstance(exc, ValidationError):  # pragma: no branch
-                message += "\n" + "\n".join(force_text(err) for err in exc.messages)
-            else:  # pragma: no branch
-                message += " " + str(exc)
-            return JsonResponse({"success": False, "errorMessage": message}, status=400)
+        return _handle_or_return_error(self._handle_create, request, _("Could not create order:"))
+
+
+def _handle_or_return_error(func, request, error_message):
+    try:
+        return func(request)
+    except Exception as exc:
+        if isinstance(exc, ValidationError):
+            error_message += "\n" + "\n".join(force_text(err) for err in exc.messages)
+        else:
+            error_message += " " + str(exc)
```

Comment:
`" " + str(exc)` should be `" {}".format(exc)` so it won't be converted to str and back to unicode in Py2 (possibly causing UnicodeEncodeError if error message has non-ASCII characters). (This was broken in the previous version too.